### ISSUE TITLE
Fixes to diff to allow things like A[i].diff(A[j]) to return KroneckerDelta(i, j)

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#11568]
See the discussions at https://github.com/sympy/sympy/pull/10594 (and also going further back, at https://github.com/sympy/sympy/pull/7437, https://github.com/sympy/sympy/issues/7421, and https://groups.google.com/forum/#!msg/sympy/wpmhMPuS8wE/eq-m6Q5hMCkJ).

I think we should do the following:

- [ ] Add new API for expressions to tell if another, _diff_wrt expression is contained in it for the purposes of differentiation. The default implementation would use `has` (or should it be `free_symbols`?).
- [ ] Make the differentiation algorithm use _diff_wrt instead of is_Symbol checks. IIRC it gets away with symbol checks because it just replaces diff_wrt expressions with dummy symbols, but that sort of thing doesn't work in general (although I think it's what lead to the idea behind this PR). 
- [ ] Remove the hacks in diff around Indexed. Indexed (and similarly MatrixElement) should be able to get the behavior they want by implementing the API method from point 1 and setting themselves as `_diff_wrt = True`. (see https://github.com/sympy/sympy/pull/11144 and specifically [this code](https://github.com/Upabjojr/sympy/blob/524043875ff55c7973f7b9e42bd0e3d37b0fb447/sympy/core/function.py#L1108))

Some somewhat related issues:

https://github.com/sympy/sympy/issues/10150
https://github.com/sympy/sympy/issues/7027
https://github.com/sympy/sympy/issues/5858

**Comment by asmeurer** [comment]
Maybe instead of the new API, we can just remove the `has`/`free_symbols` derivative=0 short circuit if a non-Symbol `_diff_wrt` expression is used.

**Comment by asmeurer** [comment]
Some of these issues also came up at https://github.com/sympy/sympy/pull/10084 in the vector module.

**Comment by Upabjojr** [comment]
> Maybe instead of the new API, we can just remove the has/free_symbols derivative=0 short circuit if a non-Symbol _diff_wrt expression is used.

That's way easier. Just hope that all current `_eval_derivative`, `fdiff` and other stuff can correctly handle derivatives of constant expressions.

**Comment by asmeurer** [comment]
We'll have to try implementing it to know for sure if it will work, but I agree that that way is much better.

**Comment by Upabjojr** [comment]
https://github.com/sympy/sympy/pull/11570

**Comment by Upabjojr** [comment]
> Make the differentiation algorithm use _diff_wrt instead of is_Symbol checks

Or maybe not, you don't want a function to be treated as _Symbol_, I'm realizing this while playing with the code. Maybe better add `is_Symbol = True` to `MatrixElement`.

**Comment by Upabjojr** [comment]
Here my fix for `MatrixElement`: https://github.com/sympy/sympy/pull/11571

**Comment by asmeurer** [comment]
In which cases should function be treated differently from MatrixElement?

**Comment by Upabjojr** [comment]
> In which cases should function be treated differently from MatrixElement? 

Function objects need to be changed with dummies, otherwise it breaks a lot of existing code. Too large a refactory.

**Comment by asmeurer** [comment]
So should this at least be set as a TODO. The whole point of _diff_wrt is that diff is agnostic toward the type of the wrt variable. Setting `is_Symbol` on something that isn't an instance of Symbol is a hack.

**Comment by Upabjojr** [comment]
> Setting is_Symbol on something that isn't an instance of Symbol is a hack. 

It has already been done (RandomSymbol, Indexed). Those classes are supposed to have the same behaviour as symbols while manipulating expressions. Technically, all classes with the `is_Symbol = True` hack could be treated as symbols in the solvers, limits, differential equations. I think this can be the way to go.

**Comment by asmeurer** [comment]
If we do this we should document it. Every other `is_Uppercase` is a shortcut for `isinstance(Uppercase)`. 

I really think it would be better to use a lowercase signifier, though. How about `is_symbol`?

**Comment by Upabjojr** [comment]
> I really think it would be better to use a lowercase signifier, though. How about is_symbol?

+1 that's a good idea.

**Comment by Upabjojr** [comment]
I've just had a look at http://orion.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf , in the section about the derivative of the determinant.

We currently have 
```python
In [2]: M = MatrixSymbol("M", 3, 3)

In [3]: var("x")
Out[3]: x

In [4]: det(M)
Out[4]: Determinant(M)

In [5]: det(M).diff(x)
Out[5]: 0
```

How can we express the fact that a `MatrixSymbol` might depend on a variable?

**Comment by asmeurer** [comment]
Should we create a MatrixFunction and require it to be explicit?

**Comment by Upabjojr** [comment]
> Should we create a MatrixFunction and require it to be explicit?

I had some thoughts about something like TensorField for the tensor module.

My fear is that a MatrixFunction object would make it much more complicated to use. I would try to suggest to go for leaving `diff(M, x)` unevaluated, but I'm not so sure.

**Comment by asmeurer** [comment]
If it's unevaluated, how would you indicate that it doesn't depend on x to clear the derivatives in an expression?

**Comment by Upabjojr** [comment]
> If it's unevaluated, how would you indicate that it doesn't depend on x to clear the derivatives in an expression?

That's true. What about allowing a `MatrixSymbol` to take a function in place of its symbol to define the matrix function? Same argument applied to `IndexedBase`, there were people on StackOverflow asking for indexed functions.

**Comment by Upabjojr** [comment]
Or... otherwise allow `Function` to take _MatrixSymbol_, _MatrixElement_, _IndexedBase_, and _Indexed_ objects?

**Comment by asmeurer** [comment]
You mean like `Function('f')(MatrixSymbol("x", 2, 3))`? The problem there is that `f` has no shape information. I would personally assume that `f` is a scalar function (`f: M_2x3 -> C`). `MatrixFunction` would take a shape, and the output would be a matrix of that shape, and actually the input could be anything. `MatrixFunction('f', 2, 3)(1)` and `MatrixFunction('f', 2, 3)(Matrix([[1, 2], [3, 4]])` would both be allowed. 

It would effectively generalize `Function`, so for instance, existing Matrix functions like Determinant or Transpose could be changed to subclass from MatrixFunction. 

> What about allowing a MatrixSymbol to take a function in place of its symbol to define the matrix function? 

I'm not sure I follow this suggestion. You mean like `MatrixSymbol(f(x), 2, 3)`? What would that mean? 

> Same argument applied to IndexedBase, there were people on StackOverflow asking for indexed functions.

Yeah, hopefully we can converge on some consistency between matrices and tensors and indexed objects in general. I personally find it easiest to think about matrices, because they are the most developed, and have the most existing functions (like determinant and trace and so on), and plus the matrix cookbook is specifically about matrices. But we should definitely keep the more general cases in mind for whatever we implement.

**Comment by Upabjojr** [comment]
> You mean like Function('f')(MatrixSymbol("x", 2, 3))? The problem there is that f has no shape information. I would personally assume that f is a scalar function (f: M_2x3 -> C). MatrixFunction would take a shape, and the output would be a matrix of that shape, and actually the input could be anything. MatrixFunction('f', 2, 3)(1) and MatrixFunction('f', 2, 3)(Matrix([[1, 2], [3, 4]]) would both be allowed. 

I mean `Function(MatrixSymbol("x", 2, 3))`.

> I'm not sure I follow this suggestion. You mean like MatrixSymbol(f(x), 2, 3)? What would that mean?

I mean something like `MatrixSymbol(Function("f"), 2, 3)(x)`

> the matrix cookbook is specifically about matrices.

That's a limitation, `Indexed` are more a multidimensional generalization of matrices that always require to be enclosed in summations for matrix/tensor multiplication. The only difference of concept I can think of is the behavior when they're regarded as function arguments: `exp(A)` for `A` indexed symbol is what? A matrix exponential? An elementwise exponential?

> But we should definitely keep the more general cases in mind for whatever we implement. 

I would like to think of a way that would work for all in one steps.

**Comment by asmeurer** [comment]
> I mean Function(MatrixSymbol("x", 2, 3)).

That's probably find as syntactic sugar. We still need the actual class structure that that would create, though. 

> I mean something like MatrixSymbol(Function("f"), 2, 3)(x)

I find that to be more confusing, especially since `f = Function('f')` is usually written separately.

**Comment by Upabjojr** [comment]
Yeah, I'd also rather go for `Function(MatrixSymbol("x", 2, 3))`.

**Comment by Upabjojr** [comment]
> That's probably find as syntactic sugar. We still need the actual class structure that that would create, though.

It could be an subclass of function overloading __getitem__.

**Comment by asmeurer** [comment]
Actually, we may be conflating two separate things. `Function` actually works in two ways, as a superclass, and as syntatic sugar for `UndefinedFunction` (`Function('f')` creates a subclass of `AppliedUndef`, whose metaclass is `UndefinedFunction`). These are completely separate things. The former behavior (superclass) is used to make `__new__` simpler (`Function` subclasses only have to define `_eval` and `Function.__new__` handles the logic of evaluating). We really only need the undefined function behavior. So I don't think we necessarily need a full blown `MatrixFunction`. We just need a Matrix equivalent of `AppliedUndef` which acts like MatrixSymbol. 

Also, I think it would be better to skip the class design of AppliedUndef (https://github.com/sympy/sympy/issues/4787). The design of IndexedBase/Indexed is actually better, because both the function and the applied function are objects that can be used in expressions. 

So I would suggest something like

```py
class MatrixUndefinedFunction(MatrixSymbol):
    def __call__(self, expr):
        return MatrixAppliedFunction(self, expr)

class MatrixAppliedFunction(MatrixExpr):
    def __new__(cls, function, *args):
        self.function = function 
        self.args = (function, *args)
        # Or, alternately
        self.func = function
        self.args = args
```

That's just a basic mockup, but the idea is that both the matrix function `f` and `f(X)` would be MatrixExpr objects. Beyond that we can add some syntactic sugar to make creating matrix functions easier. 

I actually think that just making the matrix function object a simple subclass of `MatrixSymbol` with `__call__` should work just fine, and it would be the simplest method.

**Comment by Upabjojr** [comment]
> I think that just making the matrix function object a simple subclass of MatrixSymbol with __call__ should work just fine, and it would be the simplest method. 

+1, maybe even without subclass, just `MatrixSymbol.__call__` and `MatrixElement.__call__` to create `M(x)` and `M[i, j](x)` respectively.

Possibly the same procedure could be applied for `Indexed`.

**Comment by asmeurer** [comment]
Perhaps. I know there are a [lot of people](https://github.com/sympy/sympy/issues/3539) who don't like that `Symbol.__call__` creates a function.

**Comment by Upabjojr** [comment]
> Perhaps. I know there are a lot of people who don't like that Symbol.__call__ creates a function. 

The point is about avoiding an explosion of ad-hoc classes. Consider:

- `MatrixSymbol`
- `MatrixElement`
- `IndexedBase`
- `Indexed`

They could all potentially need their _Function_-like class counterparts. Symbols and functions are two clear concepts, but are all these other ones?

**Comment by asmeurer** [comment]
I don't feel strongly about this particular issue either way.

**Comment by Upabjojr** [comment]
We now have .expr_free_symbols . It behaves like free symbols but stops the revision as long as the arguments aren't Expr anymore.
